File: src/lib/sdr-tester.ts
import { connectSSH } from './ssh';
import { Settings, StepUpdateEvent, PrepStepEvent } from './types';
import { readFileSync } from 'fs';
import { join } from 'path';
import { spawn, ChildProcess } from 'child_process';

export interface SdrTestResult {
  operation: string;
  started_at: string;
  finished_at: string;
  result: 'pass' | 'fail';
  metrics?: {
    status: string;
    peak_freq_hz: number;
    expected_freq_hz: number;
    freq_error_hz: number;
    snr_db: number;
    snr_threshold_db: number;
    peak_power_db: number;
    noise_floor_db: number;
  };
  steps: {
    id: number;
    name: string;
    label: string;
    status: string;
    message: string | null;
    duration_s: number | null;
  }[];
}

type EmitFn = (event: string, data: StepUpdateEvent | SdrTestResult | PrepStepEvent | { error: string }) => void;

function log(msg: string, ...args: unknown[]) {
  console.log(`[sdr-tester] ${msg}`, ...args);
}

function getSdrAsset(filename: string): string {
  const paths = [
    join(process.cwd(), 'src', 'assets', 'sdr', filename),
    join(process.cwd(), 'assets', 'sdr', filename),
  ];
  for (const p of paths) {
    try {
      return readFileSync(p, 'utf-8');
    } catch { /* try next */ }
  }
  throw new Error(`SDR asset not found: ${filename}`);
}

function getSdrAssetPath(filename: string): string {
  const paths = [
    join(process.cwd(), 'src', 'assets', 'sdr', filename),
    join(process.cwd(), 'assets', 'sdr', filename),
  ];
  for (const p of paths) {
    try {
      readFileSync(p);
      return p;
    } catch { /* try next */ }
  }
  throw new Error(`SDR asset not found: ${filename}`);
}

function parseProgressLine(line: string): StepUpdateEvent | null {
  const passMatch = line.match(/\[(\d+)\/\d+\]\s+(.+?)\s+—\s+PASS\s*\((.+?)\)/);
  if (passMatch) return { stepNumber: parseInt(passMatch[1]), status: 'pass', message: passMatch[3] };

  const failMatch = line.match(/\[(\d+)\/\d+\]\s+(.+?)\s+—\s+FAIL:\s*(.*)/);
  if (failMatch) return { stepNumber: parseInt(failMatch[1]), status: 'fail', message: failMatch[3] };

  const skipMatch = line.match(/\[(\d+)\/\d+\]\s+(.+?)\s+—\s+SKIPPED/);
  if (skipMatch) return { stepNumber: parseInt(skipMatch[1]), status: 'skipped', message: 'Skipped' };

  const progMatch = line.match(/\[(\d+)\/\d+\]\s+(.+?)\.{3}/);
  if (progMatch) return { stepNumber: parseInt(progMatch[1]), status: 'in_progress', message: progMatch[2].trim() };

  return null;
}

function startLocalTx(durationSeconds: number): { process: ChildProcess; kill: () => void } {
  const configPath = getSdrAssetPath('config.py');
  const txPath = getSdrAssetPath('tx_tone.py');
  const dir = join(txPath, '..');

  log('Starting local TX transmitter for %ds from %s', durationSeconds, dir);

  const proc = spawn('python3', [txPath], {
    cwd: dir,
    env: { ...process.env, PYTHONUNBUFFERED: '1' },
    stdio: ['pipe', 'pipe', 'pipe'],
  });

  proc.stdout?.on('data', (d: Buffer) => log('[TX stdout] %s', d.toString().trim()));
  proc.stderr?.on('data', (d: Buffer) => log('[TX stderr] %s', d.toString().trim()));
  proc.on('error', (err) => log('[TX error] %s', err.message));

  // Auto-kill after duration + buffer
  const timer = setTimeout(() => {
    log('TX duration elapsed, sending SIGINT');
    proc.kill('SIGINT');
  }, (durationSeconds + 1) * 1000);

  return {
    process: proc,
    kill: () => {
      clearTimeout(timer);
      if (!proc.killed) {
        proc.kill('SIGINT');
        // Force kill after 3s
        setTimeout(() => {
          if (!proc.killed) proc.kill('SIGKILL');
        }, 3000);
      }
    },
  };
}

export async function runSdrTest(
  serialNumber: string,
  settings: Settings,
  emit: EmitFn
): Promise<SdrTestResult> {
  let conn;
  let txHandle: { process: ChildProcess; kill: () => void } | null = null;

  log('Starting SDR test for T3S-%s (host: %s)', serialNumber, settings.deviceIp);

  try {
    // === PREP: Check desktop SDR ===
    emit('prep_step', { stepId: 'check_desktop_sdr', status: 'in_progress', message: 'Checking desktop SDR...' });
    try {
      getSdrAssetPath('tx_tone.py');
      getSdrAssetPath('config.py');
      log('TX scripts found');
      emit('prep_step', { stepId: 'check_desktop_sdr', status: 'pass', message: 'Desktop SDR scripts ready' });
    } catch (err) {
      const msg = err instanceof Error ? err.message : 'Unknown error';
      emit('prep_step', { stepId: 'check_desktop_sdr', status: 'fail', message: msg });
      throw new Error(msg);
    }

    // === PREP: Upload test scripts to Pi ===
    emit('prep_step', { stepId: 'upload_test_scripts', status: 'in_progress', message: 'Connecting to device...' });

    conn = await connectSSH({
      host: settings.deviceIp,
      username: settings.sshUsername,
      password: settings.sshPassword,
      timeout: 10000,
    });
    log('SSH connected');

    // Create directory and upload scripts
    await conn.exec('mkdir -p /tmp/sdr');

    const configPy = getSdrAsset('config.py');
    const rxTonePy = getSdrAsset('rx_tone.py');
    const testSh = getSdrAsset('test.sh');

    await conn.uploadFile(configPy, '/tmp/sdr/config.py');
    await conn.uploadFile(rxTonePy, '/tmp/sdr/rx_tone.py');
    await conn.uploadFile(testSh, '/tmp/sdr/test.sh');

    log('Uploaded SDR test scripts to Pi');
    emit('prep_step', { stepId: 'upload_test_scripts', status: 'pass', message: 'Test scripts uploaded' });

    // === PREP: Start desktop transmitter ===
    emit('prep_step', { stepId: 'start_transmitter', status: 'in_progress', message: 'Starting desktop transmitter...' });

    const captureDuration = 5; // seconds
    txHandle = startLocalTx(captureDuration);

    // Wait for TX to initialize
    await new Promise(resolve => setTimeout(resolve, 1500));
    log('TX transmitter started, waiting 1.5s for init');
    emit('prep_step', { stepId: 'start_transmitter', status: 'pass', message: 'Transmitter active' });

    // === RUN: Execute test.sh on Pi ===
    log('Running test.sh on Pi...');

    const command = `bash /tmp/sdr/test.sh --duration ${captureDuration} --json 2>&1`;

    let finalJson: SdrTestResult | null = null;
    let jsonBuffer = '';
    let collectingJson = false;
    let allOutput = '';
    const stepTimers = new Map<number, number>();

    const fixJson = (str: string): string => str.replace(/:(\.\d)/g, ':0$1');

    const tryParseJson = (str: string): SdrTestResult | null => {
      try {
        const parsed = JSON.parse(fixJson(str));
        if (parsed && typeof parsed === 'object' && 'operation' in parsed) {
          return parsed as SdrTestResult;
        }
      } catch { /* not valid JSON */ }
      return null;
    };

    const processData = (data: string) => {
      allOutput += data;

      if (collectingJson) {
        jsonBuffer += data;
        const parsed = tryParseJson(jsonBuffer);
        if (parsed) {
          finalJson = parsed;
          collectingJson = false;
          log('Received test JSON result: %s', finalJson.result);
          return;
        }
        const lastBrace = jsonBuffer.lastIndexOf('}');
        if (lastBrace > 0) {
          const parsed2 = tryParseJson(jsonBuffer.slice(0, lastBrace + 1));
          if (parsed2) {
            finalJson = parsed2;
            collectingJson = false;
            log('Received test JSON result (trimmed): %s', finalJson.result);
            return;
          }
        }
        return;
      }

      for (const rawLine of data.split('\n')) {
        const trimmed = rawLine.trim();
        if (!trimmed) continue;

        if (trimmed.includes('{') && trimmed.includes('"operation"')) {
          const jsonStart = trimmed.indexOf('{');
          const candidate = trimmed.slice(jsonStart);
          const parsed = tryParseJson(candidate);
          if (parsed) {
            finalJson = parsed;
            log('Received test JSON result: %s', finalJson.result);
            continue;
          } else {
            collectingJson = true;
            jsonBuffer = candidate;
            continue;
          }
        }

        const update = parseProgressLine(trimmed);
        if (update) {
          log('Step %d: %s — %s', update.stepNumber, update.message, update.status);
          if (update.status === 'in_progress') {
            stepTimers.set(update.stepNumber, Date.now());
          } else if (update.status === 'pass' || update.status === 'fail') {
            const startTime = stepTimers.get(update.stepNumber);
            if (startTime) update.duration = (Date.now() - startTime) / 1000;
          }
          emit('step_update', update);
        } else {
          log('SSH output: %s', trimmed);
        }
      }
    };

    const timeoutMs = 60 * 1000; // 1 minute for SDR test
    const timeoutPromise = new Promise<never>((_, reject) => {
      setTimeout(() => reject(new Error('SDR test timed out after 60 seconds')), timeoutMs);
    });

    const testPromise = conn.execStream(
      command,
      (data: string) => processData(data),
      (data: string) => processData(data)
    );

    const exitCode = await Promise.race([testPromise, timeoutPromise]);
    log('test.sh exited with code: %d', exitCode);

    // Stop TX
    txHandle.kill();
    txHandle = null;

    // Fallback JSON extraction
    if (!finalJson) {
      for (const line of allOutput.split('\n')) {
        const trimmed = line.trim();
        if (trimmed.startsWith('{') && trimmed.includes('"operation"')) {
          const parsed = tryParseJson(trimmed);
          if (parsed) { finalJson = parsed; break; }
        }
      }
    }

    if (finalJson) {
      emit('test_complete', finalJson);
      return finalJson;
    }

    log('No JSON result from test.sh');
    const result: SdrTestResult = {
      operation: 'sdr_test',
      started_at: new Date().toISOString(),
      finished_at: new Date().toISOString(),
      result: 'fail',
      steps: [],
    };
    emit('test_complete', result);
    return result;

  } catch (err) {
    const message = err instanceof Error ? err.message : 'Unknown error';
    log('SDR test failed: %s', message);

    let errorMsg = message;
    if (message.includes('ECONNREFUSED') || message.includes('ETIMEDOUT') || message.includes('Timed out')) {
      errorMsg = `Cannot reach device at ${settings.deviceIp} — check Ethernet cable`;
    } else if (message.includes('Authentication')) {
      errorMsg = 'Authentication failed — check credentials in Settings';
    }

    emit('test_error', { error: errorMsg });
    throw new Error(errorMsg);
  } finally {
    txHandle?.kill();
    conn?.close();
    log('Cleanup complete');
  }
}

File: backend/app/services/sdr_tester.py
import json
import logging
import os
import shutil
import subprocess
import time
from pathlib import Path
from typing import Callable

from ..services import ssh_service
from ..utils.progress_parser import OutputProcessor
from ..utils.error_handler import (
    get_operator_message, get_connection_message, diagnose_test_result,
)
from ..utils.operation_logger import write_operation_log

logger = logging.getLogger(__name__)

ASSETS_DIR = Path(__file__).parent.parent / "assets"
SDR_DIR = ASSETS_DIR / "sdr"


def _find_uhd_images_dir() -> str:
    """Find UHD images directory on this machine."""
    for base in ["/usr/share/uhd", "/usr/local/share/uhd", "/opt/uhd/share/uhd"]:
        images = os.path.join(base, "images")
        if os.path.isdir(images):
            return images
    return ""


def _load_config(path: Path) -> dict:
    """Load a JSON config file, return empty dict on error."""
    try:
        return json.loads(path.read_text())
    except (FileNotFoundError, json.JSONDecodeError) as e:
        logger.warning("Failed to load config %s: %s", path, e)
        return {}


def _resolve_config_path(name: str) -> Path:
    """Resolve user-writable config path, falling back to shipped default."""
    user_cfg = Path.home() / ".t3s-installer" / f"{name}_test_config.json"
    default_cfg = SDR_DIR / f"{name}_test_config.json"
    if user_cfg.exists():
        return user_cfg
    return default_cfg


def _log_config(config: dict, source: Path, label: str):
    """Log key config values for debugging."""
    logger.info(
        "%s config loaded from %s: center_freq=%.0f, snr_threshold=%.1f, "
        "freq_tolerance=%.0f, search_bw=%.0f, tx_gain=%.0f, rx_gain=%.0f",
        label, source,
        config.get("center_freq_hz", 0),
        config.get("snr_threshold_db", 0),
        config.get("freq_tolerance_hz", 0),
        config.get("search_bandwidth_hz", 0),
        config.get("tx_gain", 0),
        config.get("rx_gain", 0),
    )


def _log_test_summary(serial: str, result: dict, diagnosis: dict | None, config_source: Path):
    """Log a single structured test summary."""
    metrics = result.get("metrics", {})
    lines = [f"SDR test complete: serial={serial}, result={result.get('result', '?')}, config={config_source}"]

    for ch_key in ["channel_a", "channel_b"]:
        ch = metrics.get(ch_key)
        if ch:
            snr_ok = "PASS" if ch.get("snr_db", 0) >= ch.get("snr_threshold_db", 0) else "FAIL"
            freq_ok = "PASS" if ch.get("freq_error_hz", float("inf")) <= ch.get("snr_threshold_db", 0) else "?"
            lines.append(
                f"  {ch_key}: snr={ch.get('snr_db', '?')}dB (threshold={ch.get('snr_threshold_db', '?')} {snr_ok}), "
                f"freq_err={ch.get('freq_error_hz', '?')}Hz, peak={ch.get('peak_freq_hz', '?')}Hz, status={ch.get('status', '?')}"
            )

    if diagnosis:
        lines.append(f"  diagnosis: {diagnosis.get('failure_type', '?')}, is_config_issue={diagnosis.get('is_config_issue', False)}")

    logger.info("\n".join(lines))


def run_sdr_test(serial_number: str, settings, emit: Callable, dual_channel: bool = True):
    """Run SDR validation test: TX on this desktop, RX on Pi."""
    conn = None
    tx_proc = None
    num_channels = 2 if dual_channel else 1
    config_path = _resolve_config_path("sdr")
    config = _load_config(config_path)

    logger.info("Starting SDR test for %s (device: %s, channels: %d)",
                serial_number, settings.device_ip, num_channels)
    _log_config(config, config_path, "SDR test")

    try:
        # === PREP: Check desktop SDR ===
        emit("prep_step", {"step_id": "check_desktop_sdr", "status": "in_progress", "message": "Checking desktop SDR..."})

        uhd_path = shutil.which("uhd_find_devices")
        if not uhd_path:
            emit("prep_step", {"step_id": "check_desktop_sdr", "status": "fail",
                               "message": get_operator_message("sdr_test", "init_receiver", "uhd_not_found")})
            raise RuntimeError("UHD tools not installed on this machine")

        uhd_images = _find_uhd_images_dir()
        env = os.environ.copy()
        if uhd_images:
            env["UHD_IMAGES_DIR"] = uhd_images

        try:
            result = subprocess.run(["uhd_find_devices"], capture_output=True, text=True, timeout=30, env=env)
            output = result.stdout + result.stderr
            logger.info("uhd_find_devices: %s", output.strip()[:200])
        except subprocess.TimeoutExpired:
            emit("prep_step", {"step_id": "check_desktop_sdr", "status": "fail",
                               "message": get_operator_message("sdr_test", "init_receiver", "timeout")})
            raise RuntimeError("uhd_find_devices timed out")

        if "type: b200" not in output and "product: B210" not in output:
            emit("prep_step", {"step_id": "check_desktop_sdr", "status": "fail",
                               "message": get_operator_message("sdr_test", "init_receiver", "no_b210")})
            raise RuntimeError("No B210 SDR detected on this machine")

        emit("prep_step", {"step_id": "check_desktop_sdr", "status": "pass", "message": "Desktop SDR ready"})

        # === PREP: Upload test scripts to Pi ===
        emit("prep_step", {"step_id": "upload_test_scripts", "status": "in_progress", "message": "Connecting to device..."})

        conn = ssh_service.connect(settings.device_ip, settings.ssh_username, settings.ssh_password)
        logger.info("Connected to Pi")

        try:
            conn.exec_command("mkdir -p /tmp/sdr")
            conn.upload_file((SDR_DIR / "config.py").read_text(), "/tmp/sdr/config.py")
            conn.upload_file((SDR_DIR / "rx_tone.py").read_text(), "/tmp/sdr/rx_tone.py")
            conn.upload_file((SDR_DIR / "test.sh").read_text(), "/tmp/sdr/test.sh")
            conn.upload_file(config_path.read_text(), "/tmp/sdr/sdr_test_config.json")
            logger.info("Uploaded SDR test scripts to Pi (config: %s)", config_path)
        except Exception as e:
            logger.error("Failed to upload test scripts: %s", e)
            emit("prep_step", {"step_id": "upload_test_scripts", "status": "fail",
                               "message": get_operator_message("sdr_test", "run_test", "upload_failed")})
            raise RuntimeError(f"Failed to upload test scripts: {e}")

        emit("prep_step", {"step_id": "upload_test_scripts", "status": "pass", "message": "Test scripts uploaded"})

        # === PREP: Start TX locally ===
        emit("prep_step", {"step_id": "start_transmitter", "status": "in_progress", "message": "Starting transmitter..."})

        config_file = str(config_path)
        capture_duration = config.get("capture_duration_s", 5)
        tx_script = str(SDR_DIR / "tx_tone.py")

        tx_log = open("/tmp/t3s-tx-stderr.log", "w")
        try:
            tx_proc = subprocess.Popen(
                ["python3", tx_script, "--channels", str(num_channels), "--config", config_file],
                cwd=str(SDR_DIR),
                env=env,
                stdout=subprocess.DEVNULL,
                stderr=tx_log,
            )
        except Exception as e:
            tx_log.close()
            logger.error("Failed to start TX process: %s", e)
            emit("prep_step", {"step_id": "start_transmitter", "status": "fail",
                               "message": get_operator_message("sdr_test", "run_test", "tx_died")})
            raise RuntimeError(f"Failed to start TX: {e}")

        logger.info("TX started (PID: %d, channels: %d)", tx_proc.pid, num_channels)

        # Wait for TX to initialize
        init_wait = config.get("tx_init_wait_dual_s", 8) if dual_channel else config.get("tx_init_wait_s", 5)
        time.sleep(init_wait)

        if tx_proc.poll() is not None:
            tx_log.close()
            # Read TX stderr for diagnostics
            try:
                tx_stderr = Path("/tmp/t3s-tx-stderr.log").read_text()
                logger.error("TX stderr: %s", tx_stderr[:500])
            except Exception:
                pass
            emit("prep_step", {"step_id": "start_transmitter", "status": "fail",
                               "message": get_operator_message("sdr_test", "run_test", "tx_died")})
            raise RuntimeError("TX process died during initialization")

        ch_label = "double canal" if dual_channel else "canal unique"
        emit("prep_step", {"step_id": "start_transmitter", "status": "pass", "message": f"Émetteur actif ({ch_label})"})

        # === RUN: Execute test.sh on Pi ===
        logger.info("Running test.sh on Pi")

        command = f"bash /tmp/sdr/test.sh --duration {int(capture_duration)} --channels {num_channels} --config /tmp/sdr/sdr_test_config.json --json 2>&1"
        processor = OutputProcessor()

        def on_output(data: str):
            events = processor.process_data(data)
            for event in events:
                emit(event["type"], event["data"])

        try:
            exit_code = conn.exec_stream(command, on_output, timeout=int(config.get("rx_init_timeout_s", 120)))
        except Exception as e:
            logger.error("test.sh stream failed: %s", e)
            emit("test_error", {"error": str(e),
                                "operator_message": get_operator_message("sdr_test", "run_test", "streaming_timeout")})
            raise
        logger.info("test.sh exited with code: %d", exit_code)

        # Stop TX
        if tx_proc and tx_proc.poll() is None:
            tx_proc.terminate()
            try:
                tx_proc.wait(timeout=5)
            except subprocess.TimeoutExpired:
                tx_proc.kill()
            logger.info("TX stopped")
        tx_log.close()

        test_result = processor.extract_json_fallback()
        if test_result:
            # Diagnose and enrich failed results
            if test_result.get("result") == "fail":
                diagnosis = diagnose_test_result(test_result.get("metrics", {}), "sdr_test")
                test_result["diagnosis"] = diagnosis
                logger.warning("SDR test FAILED: %s", diagnosis.get("failure_type"))

                # Enrich steps with specific operator messages
                for step in test_result.get("steps", []):
                    if step.get("status") == "fail" and step.get("name") == "validate_results":
                        step["operator_message"] = diagnosis["operator_message"]
                    elif step.get("status") == "fail":
                        step["operator_message"] = get_operator_message("sdr_test", step.get("name", ""), "fail")

            _log_test_summary(serial_number, test_result, test_result.get("diagnosis"), config_path)
            write_operation_log(
                operation="sdr-test", serial=serial_number, result=test_result.get("result", "fail"),
                config=config, metrics=test_result.get("metrics"),
                diagnosis=test_result.get("diagnosis"), steps=test_result.get("steps"),
            )
            emit("test_complete", test_result)
            return test_result

        logger.warning("No JSON result from test.sh")
        fail_result = {
            "operation": "sdr_test",
            "result": "fail",
            "steps": [],
            "operator_message": get_operator_message("sdr_test", "run_test", "no_json"),
        }
        write_operation_log(
            operation="sdr-test", serial=serial_number, result="fail",
            config=config, error="No JSON result from test.sh",
        )
        emit("test_complete", fail_result)
        return fail_result

    except Exception as e:
        msg = str(e)
        logger.error("SDR test failed: %s", msg)
        if "timed out" in msg.lower() or "refused" in msg.lower():
            operator_msg = get_connection_message("unreachable")
        elif "authentication" in msg.lower():
            operator_msg = get_connection_message("auth_failed")
        elif "No B210" in msg:
            operator_msg = get_operator_message("sdr_test", "init_receiver", "no_b210")
        elif "UHD" in msg:
            operator_msg = get_operator_message("sdr_test", "init_receiver", "uhd_not_found")
        elif "upload" in msg.lower():
            operator_msg = get_operator_message("sdr_test", "run_test", "upload_failed")
        elif "TX" in msg:
            operator_msg = get_operator_message("sdr_test", "run_test", "tx_died")
        else:
            operator_msg = get_operator_message("sdr_test", "run_test", "fail")
        write_operation_log(
            operation="sdr-test", serial=serial_number, result="fail",
            config=config, error=msg,
        )
        emit("test_error", {"error": msg, "operator_message": operator_msg})
        raise
    finally:
        if tx_proc and tx_proc.poll() is None:
            tx_proc.terminate()
            try:
                tx_proc.wait(timeout=3)
            except subprocess.TimeoutExpired:
                tx_proc.kill()
        if conn:
            conn.close()
        logger.info("Cleanup complete")
